zuul-core/src/main/java/com/netflix/zuul/netty/timeouts/HttpHeadersTimeoutHandler.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2018 Netflix, Inc.
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License");
+ *      you may not use this file except in compliance with the License.
+ *      You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *      Unless required by applicable law or agreed to in writing, software
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
+ */
+
+package com.netflix.zuul.netty.timeouts;
+
+import java.util.concurrent.TimeUnit;
+import java.util.function.BooleanSupplier;
+import java.util.function.IntSupplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.netflix.spectator.api.Counter;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.codec.http.HttpMessage;
+import io.netty.handler.timeout.ReadTimeoutException;
+import io.netty.util.AttributeKey;
+import io.netty.util.concurrent.ScheduledFuture;
+
+public class HttpHeadersTimeoutHandler {
+        private static final Logger LOG = LoggerFactory.getLogger(HttpHeadersTimeoutHandler.class);
+
+        private static final AttributeKey<ScheduledFuture<Void>> HTTP_HEADERS_READ_TIMEOUT_FUTURE =
+            AttributeKey.newInstance("httpHeadersReadTimeoutFuture");
+
+        public static class InboundHandler extends ChannelInboundHandlerAdapter {
+            private final BooleanSupplier httpHeadersReadTimeoutEnabledSupplier;
+            private final IntSupplier httpHeadersReadTimeoutSupplier;
+
+            private final Counter httpHeadersReadTimeoutCounter;
+
+            private boolean closed = false;
+
+            public InboundHandler(BooleanSupplier httpHeadersReadTimeoutEnabledSupplier, IntSupplier httpHeadersReadTimeoutSupplier, Counter httpHeadersReadTimeoutCounter) {
+                this.httpHeadersReadTimeoutEnabledSupplier = httpHeadersReadTimeoutEnabledSupplier;
+                this.httpHeadersReadTimeoutSupplier = httpHeadersReadTimeoutSupplier;
+                this.httpHeadersReadTimeoutCounter = httpHeadersReadTimeoutCounter;
+            }
+
+            @Override
+            public void channelActive(ChannelHandlerContext ctx) throws Exception {
+                try {
+                    if (!httpHeadersReadTimeoutEnabledSupplier.getAsBoolean())
+                        return;
+                    int timeout = httpHeadersReadTimeoutSupplier.getAsInt();
+                    ctx.channel().attr(HTTP_HEADERS_READ_TIMEOUT_FUTURE).set(
```

Comment:
It might be a little simpler to use a `ReadTimeoutHandler` in place of dealing with the scheduling and cancellation yourself